### PR TITLE
Implemented STU3

### PR DESCRIPTION
**Pull request by jarimayenburg**
Great library! We needed STU3 support so I used your generator to generate the STU3 model and added an STU3 client with the appropriate tests. All tests should be passing, let me know what you think.

### CODE REVIEW

**Comment by FlixCoder** [comment]
Thanks, I will need some time to look up the STU3 spec and such and check everything, as it is just my free time :D

**Comment by jarimayenburg** [comment]
Take all the time you need, we will be using our fork until this is merged anyway

**Review by FlixCoder — COMMENTED** [review]
Alright, found time quite early actually, so was way faster.

I still need to check a few things, but it already looks quite good! I mean much of it is just a copy (still need to find a way to get rid of the code duplication, any nice ideas?).

Only missing checks:
- Completeness of test examples
- Correctness of JSON schema
- Generated code matches the current generate crate
- xtask test actually runs through

So everything should be fine ^^

A few remarks I have:

**Comment by FlixCoder** [comment]
Huh, apparently you aren't using the most up to date definitions of STU3. Do you mind if I switch to the newest downloads version? Do I have access to push here? :D I would of course only add a commit so your contribution would remain the same.

There is also (also done on my local branch):
- Client needs a method `stu3` to switch the client to STU3 mode.
- OCD kicking in with STU3 not being in front of R4B in some cases :P 
- VSCode settings need to enable STU3 feature to show

File: crates/fhir-sdk/src/client/mod.rs [hunk 1]
```diff
@@ -36,13 +38,19 @@ pub struct FhirR4B;
...
 pub struct FhirR5;
+/// FHIR client version to use: FHIR STU3.
+#[derive(Debug)]
+pub struct FhirStu3;
 
 #[cfg(feature = "r5")]
 /// Default client version.
 type DefaultVersion = FhirR5;
-#[cfg(all(not(feature = "r5"), feature = "r4b"))]
+#[cfg(all(not(feature = "r5"), not(feature = "stu3"), feature = "r4b"))]
```
[the remark this answers is not in the document]
Comment (jarimayenburg) [reply crates/fhir-sdk/src/client/mod.rs:48]: Sounds good!

File: crates/generate/src/generate/gen_types.rs [hunk 2]
```diff
@@ -261,7 +261,11 @@ fn generate_choice_field(
 	let enum_type = format_ident!("{type_ident}{}", field.name.replace("[x]", "").to_pascal_case());
 	let enum_doc = format!(" Choice of types for the {} field in {type_ident}", field.name);
 
-	let variants = field.types.iter().map(|ty| {
+	let mut types = field.types.clone();
+
+	types.dedup();
```
[the remark this answers is not in the document]
Comment (jarimayenburg) [reply crates/generate/src/generate/gen_types.rs:266]: You are right, that would be much better. Let me fix that.

File: crates/generate/Cargo.toml [hunk 3]
```diff
@@ -6,7 +6,7 @@ version = "0.1.0"
 [dependencies]
 Inflector = "0.11.4"
 anyhow = "1.0.70"
-fhir-model = {version = "0.6.0", default-features = false, features = ["r4b", "r5"]}
+fhir-model = {path = "../fhir-model", default-features = false, features = ["stu3", "r4b", "r5"]}
```
[the remark this answers is not in the document]
Comment (FlixCoder) [reply crates/generate/Cargo.toml:9]: No lets keep it, I will do it once I released a new version with stu3

File: crates/generate/src/generate/gen_types.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jarimayenburg) [reply crates/generate/src/generate/gen_types.rs]: I moved it now. The reason that this change is needed is that in STU3, `ElementDefinition.type.targetProfile` has cardinality `0..1` in STU3, while in R4 and up `ElementDefinition.type.targetProfile` has cardinality `0..*`. So in STU3, an element that can be of type "Reference" with multiple allowed target profiles is represented in `ElementDefinition.type` as multiple entries with the same `code` "Reference". Without this deduplication, the code generator would generate enums with multiple variants called "Reference".

**Comment by jarimayenburg** [comment]
> Huh, apparently you aren't using the most up to date definitions of STU3. Do you mind if I switch to the newest downloads version? Do I have access to push here? :D I would of course only add a commit so your contribution would remain the same.

The version of STU3 I used was 3.0.2, released on Feb 21, 2017. That should be the latest one right? Like you, I had to edit a bunch of things in the spec JSON files to get it working. Some things I changed:
1. Added `experimental`, `version` and `fhirVersion` fields to StructureDefinitions and CodeSystems wherever they were missing.
2. Added `code` fields to ElementDefinitions of primitive datatypes that only had a `_code` extension.

> * Client needs a method `stu3` to switch the client to STU3 mode.
Let me add that!

> * OCD kicking in with STU3 not being in front of R4B in some cases :P
Haha I had the same OCD kicking in. Problem is that `rustfmt` orders modules alphabetically, so `stu3` will land after `r4b` and `r5`. Feel free to order it however you'd like when it's merged though.

> * VSCode settings need to enable STU3 feature to show
Whoops, I don't use VSCode so I missed that. Added it now.

File: crates/generate/src/generate/gen_types.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (FlixCoder) [reply crates/generate/src/generate/gen_types.rs]: Ah, ok, thanks

**Comment by FlixCoder** [comment]
> he version of STU3 I used was 3.0.2, released on Feb 21, 2017. That should be the latest one right?

Hmm the `version.info` file says it is 3.0.1 :D And the newest version is a build of `3.0.2` from 2019 apparently.

> Haha I had the same OCD kicking in. Problem is that rustfmt orders modules alphabetically, so stu3 will land after r4b and r5. Feel free to order it however you'd like when it's merged though.

No I am using rustfmt too ^^ So I will need to live with it :D


Alright, I think that's it, I need to take a final look and then I will merge it, thank you very much!

**Comment by jarimayenburg** [comment]
> > he version of STU3 I used was 3.0.2, released on Feb 21, 2017. That should be the latest one right?
> 
> Hmm the `version.info` file says it is 3.0.1 :D And the newest version is a build of `3.0.2` from 2019 apparently.

On the version history page (https://hl7.org/fhir/history.html), in that big table, I think the correct latest version is called "3.0.0" and released on Mar 21, 2017. While at the top of the page they refer to Release 3 STU as 3.0.2, released on Feb 21, 2017. Not sure how a newer version 3.0.2 has an earlier release date than the older version 3.0.0, but here we are.

Anyway, I used the link to 3.0.2 in the "major milestones" to get to the final STU3 spec and got the specification from the download page of that version: https://hl7.org/fhir/stu3/downloads.html. I can't imagine that that is an older version.

**Comment by FlixCoder** [comment]
Alright, then it should be the newest :D Weird. No, totally fine then, thanks!

**Pull request merged**